CPU tensors

kohr-h opened the pull request:
### TODOs

- [x] Implement basic structures for tensors and tensor spaces
- [x] Replicate functionality of `NtuplesBase`, `FnBase`, `NumpyNtuples` and `NumpyFn` in the corresponding tensor classes
- [x] Remove `Ntuples` and related stuff
- [x] Add former `Ntuples` and the likes as special case to tensor classes
- [x] Change code that was restricted to 1D data containers
- [x] Port unit tests
- [x] Update documentation
- [x] ~~Write extra documentation on tensors~~ Should be part of #908
- [x] Port `show` methods (1D from `Ntuples`, 2D from `DiscreteLp`)
- [x] Implement tensor-valued functions, see #908 
- [x] Fix interpolation from array of points, related to #908 
- [x] Implement indexing of `FunctionSpace` for input and output dimensions 
- [x] Implement indexing of `DiscreteLp` along input dimensions (output not yet applicable since tensor-valued stuff remains TODO)
- [x] Implement `__array_ufunc__` interface on `Tensor` and `NumpyTensor` (the latter only differs in weight propagation)
- [x] Implement `__array_ufunc__` interface on `DiscreteLpElement`
- [x] Write more extensive tests for `__array_ufunc__` interfaces (bad input etc.)
  - [x] For `NumpyTensor`
  - [x] For `DiscreteLpElement`
- [x] Switch Numpy 1.13 on again
  - [x] in `requirements.txt`
  - [x] in `conda/meta.yaml`
  - [x] in `.travis.yml`
- [x] Fix handling of weighting in `DiscreteLp`. Now that `NoWeighting` goes away, we are left with only one option. We should probably default to `None` instead and allow arrays as weightings that just propagate to the `dspace` constructor. 
- [x] Add documentation of `__array_ufunc__` interface
- [x] Check all examples
- [x] Check `contrib` modules
- [x] Check boolean array and index array indexing
- [x] Fix Huber functional large-scale test
- [x] Check if all recent `npy_ntuples` and `base_ntuples` changes are being used, same for tests.

kohr-h commented:
The `__array_ufunc__` interface is now finished for both `Tensor` and `DiscreteLpElement` objects. It is fully functional for both and has extensive tests for the former, while the latter needs some more tests.

I really like the interface, it's reasonable and quite easy to understand (kudos to the Numpy guys for that). Our code does have its complexity, but most of it is frankly just from what we add on top, i.e., there's a bit of unwrapping and re-wrapping, and in particular finding the correct space for a result of, e.g., `ufunc.reduce` is not completely trivial, especially for `DiscreteLpElement`.

So I'd say the last big chunk of work here is done, this is really close to ready for merge. Only a few minor TODOs left, and I guess also a final review.

kohr-h commented on odl/discr/lp_discr.py:
```diff
@@ -686,65 +675,430 @@ def __setitem__(self, indices, values):
+        Numpy ufuncs`_.
+
+        .. note::
+            This basic implementation casts inputs and
```
This note is actually obsolete

kohr-h commented on odl/discr/lp_discr.py:
```diff
@@ -686,65 +675,430 @@ def __setitem__(self, indices, values):
+            the resulting array is wrapped in a space of the same
+            type as ``self.space``, however **only** using the minimal
+            set of parameters ``size`` and ``dtype``. If more properties
+            are supposed to be propagated, this method must be overridden.
```
Needs update

kohr-h commented on odl/discr/lp_discr.py:
```diff
@@ -686,65 +675,430 @@ def __setitem__(self, indices, values):
+            ``'__call__'``, ``'accumulate'``, ``'at'``, ``'outer'``,
+            ``'reduce'``
+
+        input1, ..., inputN:
```
missing space before colon

kohr-h commented on odl/discr/lp_discr.py:
```diff
@@ -686,65 +675,430 @@ def __setitem__(self, indices, values):
+            was given, the result is an `NtuplesBaseVector` or a tuple
+            of such, depending on the number of outputs of ``ufunc``.
+            If ``out`` was provided, the returned object or sequence members
+            refer(s) to ``out``.
```
Needs update

kohr-h commented on odl/discr/lp_discr.py:
```diff
@@ -686,65 +675,430 @@ def __setitem__(self, indices, values):
+                    # so we need to do it manually here.
+                    if all(isinstance(inp.space.weighting, ConstWeighting)
+                           for inp in (inp1, inp2)):
+                        weighting = part.cell_volume
```
Do this only if the old constants were also the cell volumes

kohr-h commented on odl/discr/lp_discr.py:
```diff
@@ -434,15 +409,99 @@ def _dist(self, x, y):
+                    else:
+                        newshape = tuple(space.shape[int(i)] for i in indices)
+
+                    weighting = part.cell_volume
```
only if the old weighting was also the cell volume

kohr-h commented on odl/ufunc_ops/ufunc_ops.py:
```diff
@@ -24,10 +24,75 @@
 
 __all__ = ()
 
+# TODO: these are implementation-dependent, add to `entry_points`?
```
Actually not as used here, remove comment

kohr-h commented on odl/util/utility.py:
```diff
@@ -35,131 +37,122 @@
-        return array1d_str(array)
+    a = np.asarray(a)
+    max_shape = tuple(a.shape[i] if a.shape[i] < nprint else nprint
+                      for i in range(a.ndim))
```
loop over `a.shape`

kohr-h commented on odl/util/utility.py:
```diff
@@ -171,6 +164,8 @@ def dtype_repr(dtype):
     elif dtype == np.dtype(complex):
         return "'complex'"
+    elif dtype.shape:
+        return "('{}', {})".format(dtype.base, dtype.shape)
```
Use `!s` conversion

adler-j commented:
Should I give this a final review then perhaps?

kohr-h commented:
> Should I give this a final review then perhaps?

Would be very much appreciated.

kohr-h commented:
I fixed also the largescale tests. In the ray transform largescale tests, the master branch currently has a lot of failures, perhaps due to too strict proximity checks. This branch makes the checks less strict and brings the failures down, but we should keep an eye on this.

adler-j commented:
Just commenting to say I'm about halfway through a review of this. Will take a few more days sadly but I'm quite confident I'll mark it "approved" since most things so far are very minor!

adler-j commented:
Chill with the rebase meanwhile to not screw up my comments in waiting!

kohr-h commented:
No worries, I'll resist my urge to rebase. (It's a larger change anyway due to the changes in the `entry_points` machinery all over the place).

adler-j commented on odl/space/space_utils.py:
```diff
@@ -51,200 +55,245 @@ def vector(array, dtype=None, impl='numpy'):
+    --------
+    Space of real 3-tuples with ``float32`` entries:
+
+    >>> odl.rn(3, dtype='float32')
```
Imo first example should be without explicit dtype

adler-j commented on odl/space/weighting.py:
```diff
@@ -453,7 +407,7 @@ def equiv(self, other):
 
         elif isinstance(other, ConstWeighting):
-            if self.matrix_issparse:
+            if scipy.sparse.isspmatrix(self.matrix):
```
Not quite sure why we did this change?

kohr-h commented:
> So! Finally finished reviewing the actual code changes. Will review the tests next.
> 
> Looking very good.
> 
> So basically fix this, rebase on master and I'll have a look at the tests after that.

I'll get right at it!

kohr-h commented on doc/source/guide/numpy_guide.rst:
```diff
@@ -70,14 +71,38 @@ For other arbitrary functions, ODL vector space elements are generally accepted
---------------------
-The fact that the ``x.ufuncs.negative()`` interface is needed is a known issue with NumPy and a `fix is underway
-<http://docs.scipy.org/doc/numpy-dev/neps/ufunc-overrides.html>`_. As of April 2016, this is not yet available.
+A known limitation of this approach is that ODL objects are not accepted as ``out`` parameters of Numpy functions.
```
yes

kohr-h commented on odl/discr/lp_discr.py:
```diff
@@ -686,65 +684,430 @@ def __setitem__(self, indices, values):
-        >>> x.ufuncs.add(3)
-        uniform_discr(0.0, 1.0, 2).element([4.0, 1.0])
+        .. _interface documentation:
+           https://github.com/charris/numpy/blob/master/doc/source/reference/\
```
Great catch, I picked it from the release docs. The proper typeset version of that is [here](https://docs.scipy.org/doc/numpy/reference/arrays.classes.html#numpy.class.__array_ufunc__).

kohr-h commented on odl/discr/discr_mappings.py:
```diff
@@ -309,60 +297,52 @@ class NearestInterpolation(FunctionSetMapping):
-        fset : `FunctionSet`
-            The undiscretized (abstract) set of functions to be
-            discretized. The function domain must provide a
+        fspace : `FunctionSpace`
```
That's perfectly fine. I also feel the small things won't ruin the PR if they slip in unfixed. We can still improve things later on if the big stuff is solid.

kohr-h commented on odl/deform/linearized.py:
```diff
@@ -76,8 +76,10 @@ def _linear_deform(template, displacement, out=None):
+        image_pts[:, i] += vi.asarray().ravel(order=vi.order)
+    values = template.interpolation(image_pts.T, out=out, bounds_check=False)
+    return values.reshape(displacement.space[0].shape,
+                          order=displacement.space[0].order)
```
Reason?
Also, `DiscreteLp` has `default_dtype` not implemented since it depends on `impl`.

Perhaps `values.reshape(template.space.shape, order=template.space.order)` makes the intentions more clear? (They should all be the same anyway according to the doc.)

kohr-h replied:
> Also, DiscreteLp has default_dtype not implemented since it depends on impl.

That is, it cannot be implemented as `@staticmethod` as for the other classes.

kohr-h commented on odl/discr/discr_mappings.py:
```diff
@@ -667,23 +584,29 @@ def __repr__(self):
-            sep = ', '
+        posargs = [self.range, self.grid, self.domain, schemes]
+
+        nontriv_var = None
```
Like that solution!

kohr-h commented on odl/discr/lp_discr.py:
```diff
@@ -116,43 +109,37 @@ def __init__(self, fspace, partition, dspace, exponent=2.0,
         else:
-            self.__order = str(order).upper()
+            # Got single string
+            interp, interp_in = str(interp + '').lower(), interp
```
I'll remove it, what the heck. Silly idea, particularly if not caught.

kohr-h commented on odl/discr/lp_discr.py:
```diff
@@ -434,15 +409,103 @@ def _dist(self, x, y):
+
+                if isinstance(space.weighting, NoWeighting):
+                    # Extra case since NoWeighting is a subclass of
+                    # ConstWeighting
```
Do we want to keep the singleton pattern for the `const=1` case? Does it matter in terms of efficiency?

kohr-h replied:
Related issue: This change will also wipe the `'none'` option to weighting in `uniform_discr`. I suggest we change that parameter completely to default to `None` and propagate as-is to the `dspace` constructor. That way we can easily construct arbitrarily weighted spaces.

adler-j replied:
Thought some more about this and second thought we should NOT remove `NoWeighting`. Specifically it conveys a different intention and distinguishes "accidental" no-weighting from intentional. That is, `uniform_discr(0, 5, 5)` has a weighting that is "conceptually" different from `rn(5)`, even if it practically is the same. Right?

kohr-h replied:
Well, they would still be distinguishable by `type()`. A more tricky case would be `uniform_discr([0, 0], [4, 4], (4, 4))` vs. `uniform_discr([0, 0], [4, 4], (2, 8))`. Both would result in constant weighting with constant 1.0. Hence not weighted. Basically the `space.is_weighted` would reduce to a pure convenience, or to something that allows a few optimizations.

Would anything go wrong? Since implementation-wise, also the old "no weighting" case was equivalent to constant 1, so I expect trouble only when doing things to axes, like `reduce` or `outer`, or indexing `byaxis_in`.
But I think everything is fine, since `DiscreteLp` always knows how the constant came to be, so it can distinguish between "arbitrary" constants and cell volume. For the latter, it will do the reasonable thing and re-compute the constant.

As a matter of fact, I already made most of the changes, and only one test fails that was constructed to test weighting propagation in a way that has to fail now.

Random thought: when we go about implementing lists of weights (like arrays per axis with broadcasting), we can obviously also allow constants per axis, and as an optimization multiply them together, making it equivalent to a single constant. The good thing would be that `TensorSpace` would know how to handle adding/removing axes in terms of weighting.

kohr-h commented on odl/discr/lp_discr.py:
```diff
@@ -686,65 +684,430 @@ def __setitem__(self, indices, values):
+
+        # Pull out the `tensor` attributes from DiscreteLpElement instances
+        # since we want to pass them to `self.tensor.__array_ufunc__`
+        input_tensors = tuple(
```
This step is to unroll the type hierarchy `DiscreteLpElement -> NumpyTensor -> numpy.ndarray` (in the sense of composition, not inheritance). Anything that is not converted is treated as array-like in the underlying `self.tensor.__array_ufunc__`. I'm not entirely sure what happens if we don't do this, potentially we run into an infinite loop of `__array_ufunc__` redirection. We want to get to the `numpy.ndarray.__array_ufunc__` at some point.
I'll check what happens without the conversion.

With that said, array-like input will also work just fine.

kohr-h commented on odl/discr/lp_discr.py:
```diff
@@ -686,65 +684,430 @@ def __setitem__(self, indices, values):
+                raise NotImplementedError('nout = {} not supported'
+                                          ''.format(ufunc.nout))
+
+        elif method == 'reduce' and keepdims:
```
Well, your example is for a `TensorSpace`, where this is meaningful and allowed. But what would you do with a `DiscreteLp` where you would have to collapse an axis to a single point? I don't see a meaningful way to do this.

kohr-h replied:
Part of the answer is that the result in your example is wrong, it should be `rn(1).element([3.0])`. The `keepdims=True` option means "keep axes of size 1 instead of removing them". It does **not** mean "broadcast the result to the original shape".

adler-j replied:
It seems you are right, agree with this then.

kohr-h replied:
Just for the record: for the same reason, `outer` requires two `DiscreteLpElement` objects, since otherwise we don't know how to infer the new space. A similar reason applies to `reduceat`.

kohr-h commented on odl/discr/lp_discr.py:
```diff
@@ -686,65 +684,430 @@ def __setitem__(self, indices, values):
+                elif method == 'outer':
+                    # Concatenate domains, partitions, interp, axis_labels,
+                    # and determine `dspace` from the result tensor
+                    inp1, inp2 = inputs[:2]
```
Probably never, the `self.tensor.__array_ufunc__` would raise already in that case.

adler-j replied:
Rather just unpack this with `inp1, inp2 = inputs` so we get a proper error in unexpected cases then

kohr-h commented on odl/discr/lp_discr.py:
```diff
@@ -686,65 +684,430 @@ def __setitem__(self, indices, values):
+                              inp2.space.interp_byaxis)
+                    labels = inp1.space.axis_labels + inp2.space.axis_labels
+                    if len(set(labels)) != len(labels):
+                        # Duplicates, take default
```
Fine with me

kohr-h commented on odl/discr/lp_discr.py:
```diff
@@ -686,65 +684,430 @@ def __setitem__(self, indices, values):
+                            weighting=None)
+                    elif all(isinstance(inp.space.weighting,
+                                        ConstWeighting)
+                             for inp in (inp1, inp2)):
```
Thanks. It was the only way to keep my thoughts straight while implementing this.

kohr-h commented on odl/discr/lp_discr.py:
```diff
@@ -1240,13 +1631,30 @@ def uniform_discr(min_pt, max_pt, shape, exponent=2.0, interp='nearest',
+
+            Default for ``'numpy'``: ``'float64'`` / ``'complex128'``
+
+            Default for ``'cuda'``: ``'float32'``
```
How about
```
    dtype : dtype, optional
        Data type for the discretized space. The default depends on the
        ``impl`` backend, usually it is ``'float64'`` or ``'float32'``.
```
?

adler-j replied:
I'd go for

... It is usually float64, but some backends may use other options like float32.

kohr-h commented on odl/operator/tensor_ops.py:
```diff
@@ -844,59 +879,646 @@ def inverse(self):
+
+        self.__sampling_points = _normalize_sampling_points(sampling_points,
+                                                            domain.ndim)
+        # Flatten indices during init for faster indexing later
```
Yes, it's like that in Numpy (at least in the cases I checked so far). Example:
```python
>>> arr = np.zeros((100, 100, 100))
>>> indcs_multi = []
>>> for n in arr.shape:
...     idx_arr = np.random.randint(0, n - 1, size=int(1e4))
...     indcs_multi.append(idx_arr)
>>> indcs_flat = np.ravel_multi_index(indcs_multi, arr.shape)
>>> %timeit arr[indcs_multi]
17 ms ± 38.8 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
>>> %timeit arr.ravel()[indcs_flat]
3.32 ms ± 17.2 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
>>> %timeit np.ravel_multi_index(indcs_multi, arr.shape)
6.33 ms ± 5.57 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```
So flattening + indexing in flat mode is faster than indexing directly. (Perhaps NumPy has a performance issue there).

kohr-h replied:
Made an issue https://github.com/numpy/numpy/issues/9635

kohr-h commented on odl/operator/tensor_ops.py:
```diff
@@ -689,119 +695,152 @@ def __init__(self, vfspace, weighting=None):
+
+        The operator also works on `uniform_discr` type spaces. Note,
+        however, that the ``weighting`` of the domain is propagated to
+        the range by default:
```
Right

kohr-h commented on odl/operator/tensor_ops.py:
```diff
@@ -844,59 +879,646 @@ def inverse(self):
+
+
+def _normalize_sampling_points(sampling_points, ndim):
+    """Normalize points to an ndim-long list of linear index arrays."""
```
Will do

kohr-h commented on odl/operator/tensor_ops.py:
```diff
@@ -844,59 +879,646 @@ def inverse(self):
+
+    The operator is defined by ::
+
+        SamplingOperator(f) == c * f[indices]
```
`sampling_points` it should be.

kohr-h commented on odl/operator/tensor_ops.py:
```diff
@@ -844,59 +879,646 @@ def inverse(self):
+
+    def _call(self, x):
+        """Return values at indices, possibly weighted."""
+        out = x.asarray().ravel()[self._indices_flat]
```
As written above, it's faster (for anything more than, say, 10000 sampling points).

kohr-h commented on odl/operator/tensor_ops.py:
```diff
@@ -844,59 +879,646 @@ def inverse(self):
+        return repr(self)
+
+
+class FlatteningOperator(Operator):
```
Done. #1106

kohr-h commented on odl/operator/pspace_ops.py:
```diff
@@ -1036,8 +1040,8 @@ def __init__(self, *operators, **kwargs):
         ProductSpace(rn(3), 2).element([
-            [1.0, 2.0, 3.0],
-            [8.0, 10.0, 12.0]
+            [ 1.,  2.,  3.],
```
OK 😇

kohr-h commented on odl/phantom/noise.py:
```diff
@@ -56,7 +56,7 @@ def white_noise(space, mean=0, stddev=1, seed=None):
                       for subspace in space]
         else:
-            if space.is_cn:
+            if space.is_complex_space:
```
Actually yes.

kohr-h commented on odl/phantom/noise.py:
```diff
@@ -123,19 +123,18 @@ def salt_pepper_noise(vector, fraction=0.05, salt_vs_pepper=0.5,
         to noise.
     salt_vs_pepper : float, optional
-        Relative aboundance of salt (high) vs pepper (low) noise. A high value
+        Relative abundance of salt (high) vs pepper (low) noise. A high value
```
I don't have a strong opinion, but "Operator Discreti**z**ation Library" kind of leaves no ambiguity.

kohr-h commented on odl/solvers/functional/derivatives.py:
```diff
@@ -56,15 +56,14 @@ def __init__(self, operator, point, method='forward', step=None):
-               [ 0.,  2.,  0.],
-               [ 0.,  0.,  2.]])
+        >>> hess_matrix = odl.matrix_representation(hess)
+        >>> np.allclose(hess_matrix, 2 * np.eye(3))
```
I also prefer it written out, but the problem was some of the zeros being printed as `-0.`, and I don't trust the reliability of that output. I'll check it again though.

adler-j replied:
Well allclose should handle that, right?

```python
>>> np.allclose(hessian_matrix, [[2, 0, 0], [0, 2, 0], [0, 0, 2]])
```

kohr-h commented on odl/space/base_tensors.py:
```diff
@@ -0,0 +1,1031 @@
+        # We choose this order in contrast to Numpy, since we usually want
+        # to represent discretizations of vector- or tensor-valued functions,
+        # i.e., if dtype.shape == (3,) we expect f[0] to have shape `shape`.
+        self.__shape = dtype.shape + shape
```
Yeah well I guess so 😁

adler-j replied:
has this been done?

kohr-h replied:
Now yes

kohr-h commented on odl/space/base_tensors.py:
```diff
@@ -0,0 +1,1031 @@
+        return self.__dtype
+
+    @property
+    def is_real_space(self):
```
Yes

kohr-h commented on odl/space/base_tensors.py:
```diff
@@ -0,0 +1,1031 @@
+        Integers and slices index the first axis:
+
+        >>> space = odl.rn((2, 3, 4))
+        >>> space[1]
```
Yeah, this whole thing gives me some headache. The only reason I can currently see to keep it implemented is compatibility with `ProductSpace`.
Other than that, I really have a hard time thinking of `rn((2, 3))` as a stack of 2 `rn(3)`'s. Indexing something whose constituents are indistinguishable from each other really adds very little value. That is, `space[0] == space[1]` is **always** true. What's the point then?

And the argument that `space[indices]` should be the same as `element[indices].space` doesn't really convince me either. It just means lots of extra code for little gain, since we index `element` in a lazy way anyway. So then one could go directly with `element[indices].space`.

Another reason for removing `__getitem__` on `TensorSpace` is that its big cousin `DiscreteLp` will not implement it since we have a huge ambiguity regarding **what** we want to index. For vector-valued functions, are we after the **input** or **output** dimensions?

I'm happy to hear other arguments. Currently I'm leaning towards dumping this method since "In the face of ambiguity, refuse the temptation to guess." (Zen of Python)

kohr-h replied:
Just for the record: simply deleting `__getitem__` of both `TensorSpace` and `NumpyTensorSpace` makes nothing fail except the unit tests designed to test exactly those methods. That means we don't need them.

kohr-h replied:
> Yeah, this whole thing gives me some headache. The only reason I can currently see to keep it implemented is compatibility with ProductSpace.

And also there, it's a new feature that may be worth a discussion.

adler-j replied:
I'd move this code to a gist or something and open an issue where we can discuss the future of them. Leave it out of this PR

kohr-h commented on odl/space/base_tensors.py:
```diff
@@ -0,0 +1,1031 @@
+           https://docs.scipy.org/doc/numpy/reference/generated/\
+numpy.ufunc.reduceat.html
+        """
+        # --- Process `out` --- #
```
I know, it's a HUGE pain just for a couple of different things. But the differences (allowed `out`, space wrapping) are big enough to make a helper function complex.

One possible alternative would be to run the parent method in `NumpyTensor` and wrap the result in a new space according to the wrapping rules. But I'm not sure how many lines that would really save.

adler-j replied:
Lets leave it at this then, but perhaps make a comment about this difficulty (for future maintainers)

kohr-h commented on odl/space/base_tensors.py:
```diff
@@ -0,0 +1,1031 @@
+
+        # --- Evaluate ufunc --- #
+
+        # Trivial context to be used when an output is `None`
```
I'll add a comment, the usefulness of this is rather limited I think.

kohr-h commented on odl/space/fspace.py:
```diff
@@ -6,371 +6,1277 @@
+    result = np.array(func._call_out_of_place(x, **kwargs), copy=False)
+    if result.dtype == object:
+        # Different shapes encountered, need to broadcast
+        flat_results = result.ravel()
```
I'll check, but I think the order is fixed here, namely determined by the input `x`, which is usually a meshgrid. Hence `'C'` order is correct, since it corresponds to the coordinate axes in `np.broadcast(*x)`.

kohr-h replied:
Ah no, this is entirely internal. We flatten the object array holding the results (broadcasting needed), then we broadcast the results to final shape and create a new array from those arrays. That array is still flat in the output dimensions, so we reshape to `out_shape + scalar_shape`.
As long as that last reshape uses the same order as the initial ravel, things are consistent.

kohr-h replied:
I'll add some comments, but honestly this part was a lot of trial-and-error since there is just so much stuff going on that it's hard to keep track of all the possible reshaping and broadcasting cases. I trust the unit tests here, they are fairly comprehensive.

adler-j replied:
Agree that "proof by unit test", while perhaps not satisfying, is quite OK. Add a comment and we're fine

kohr-h commented on odl/space/npy_tensors.py:
```diff
@@ -0,0 +1,2405 @@
+import scipy.linalg as linalg
+
+from odl.set.sets import RealNumbers, ComplexNumbers
+from odl.set.space import LinearSpaceTypeError
```
Wow, radical! But I like it! Let's keep it for a new issue though.

kohr-h replied:
#1107

adler-j replied:
I guess this gets done "later"?

kohr-h replied:
Yes, new issue

kohr-h commented on odl/space/npy_tensors.py:
```diff
@@ -0,0 +1,2405 @@
+        tensor_space((2, 3), 'int')
+        """
+        TensorSpace.__init__(self, shape, dtype, order)
+        if self.dtype.char not in self.available_dtypes():
```
It's the easiest way to check the scalar base dtype, while also including string data types. Here, `dtype.base` would not do the job.

adler-j replied:
Shouldn't we have something like "base dtype" to handle this?

adler-j commented on odl/space/npy_tensors.py:
```diff
@@ -0,0 +1,2405 @@
+            return self
+        elif self.space.is_complex_space:
+            # Definitely non-contiguous
+            real_space = self.space.astype(self.space.real_dtype, order='A')
```
I now remember our discussion on this, but add a comment to clarify

kohr-h replied:
We need to set `order='A'` because the array will be discontiguous.
Alternatively we could always use `order='A'` when using `.real` on a complex space. But I would find that counter-intuitive, I'd expect the order to be preserved in that operation **on spaces**.

kohr-h commented on odl/space/npy_tensors.py:
```diff
@@ -0,0 +1,2405 @@
+        weighting = kwargs.pop('weighting', None)
+        exponent = kwargs.pop('exponent', getattr(weighting, 'exponent', 2.0))
+
+        if (not is_numeric_dtype(self.dtype) and
```
That would have to be `or not all(...`, is that easier to understand?

adler-j replied:
why not `or any(x is None ...`?

kohr-h commented on odl/space/npy_tensors.py:
```diff
@@ -0,0 +1,2405 @@
+
+    def __hash__(self):
+        """Return ``hash(self)``."""
+        return hash((type(self), self.shape, self.dtype, self.order,
```
That seems to be closely related to #1028. Pickling would likely want to use the same properties. Perhaps one can even add them on a class level, like in theano.

kohr-h commented on odl/space/npy_tensors.py:
```diff
@@ -0,0 +1,2405 @@
+        if self.size != 1:
+            raise TypeError('only size-1 tensors can be converted to '
+                            'Python scalars')
+        return complex(self.data[(0,) * self.ndim])
```
The indexing operation takes out the 0-th element and then applies `complex`. I think that will work also with old Numpy since the result is a Numpy scalar.
Perhaps a simpler way to write this is `complex(self.data.ravel()[0])`.

kohr-h commented on odl/space/npy_tensors.py:
```diff
@@ -0,0 +1,2405 @@
+        `dtype`. The latter can be changed with the ``dtype`` parameter:
+
+        >>> x = r3.element([1, 2, 3])
+        >>> x.__array_ufunc__(np.add, 'accumulate', x)
```
Because Numpy <1.13 will use the old interface, and that will fail with our own container type as `out`.
This doctest is probably more for developers anyway, so I find it more useful like this when the interface is explicitly and reliably used.

adler-j commented:
How do we do here?

kohr-h commented:
This branch is fine, the only thing left to do here is fix the `super()` thing that makes the Python 2 test fail, and then run the examples and the `contrib` stuff, to make sure that everything works.

Work is underway in #739 to check the TODOs there.

kohr-h commented:
Porting the upstream fix for `real` and `imag` I again ran into the issue with enforced `order` in spaces. We had a discussion on that, but it's buried in some page on GH that won't open anymore.

As it turns out, I don't see a way to fix the issue for spaces with enforced contiguousness, that `x.real` for example doesn't work with assignment `x.real[:] = something`.
The problem here is that `x.real` is a discontiguous array by nature, and a `'C'` ordered space, for instance, makes a copy to get contiguous memory. Hence, writing to that fella won't change `x`.

So sorry for raising this issue again, but to me that's an indication that we shouldn't be enforcing contiguousness on a space level. I get the argument that some libraries need their inputs to be contiguous in some ordering, but why is that a reason to enforce it in a space? We can just as well

- have an option `order` in `asarray()` for example that gives an array with a certain ordering and/or
- make the `astype` method more general by renaming it to `as` and adding `dtype`, `order` and perhaps even `impl` to it.

Given also the fact that `'C'` and `'F'` spaces will be making lots of copies that hurt efficiency without clearly warning the user, I feel that we need to think more about this.

adler-j commented:
Well 98% of users will use `'A'` order anyway, right? And allowing `space.real_space.order != space.order` should solve the issue you gave, right?

kohr-h commented:
> Well 98% of users will use 'A' order anyway, right? And allowing space.real_space.order != space.order should solve the issue you gave, right?

Sure, but the same already applies to slicing operations, i.e., the sliced array is wrapped in an `'A'` space. In some sense that defeats the whole point of having a `'C'` or `'F'` space, no? What's the point of having a space property that gets dropped as soon as you change an array by slicing or similar? Then I'll anyway have to manually cast to contiguous when interfacing with libraries that require it.

adler-j commented:
> In some sense that defeats the whole point of having a 'C' or 'F' space, no?

No not really. The reason we have those is explicitly when we need to enforce it for some latter operator for some reason. Using it with sub-libraries would hardly be much of an issue I guess since calling them with slices is not too common.

The thing I'm worried about there is that by removing this we'll 100% break some use cases with no turning back possible. In particular, in the GPUMCI package I need to use `'F'` ordering (since CUDA uses that for its textures, which is far beyond my control), and if we removed this feature here it would be impossible for me to create a space that simply wraps an ODL object and uses it with GPUMCI. The only option would be to ask for a copy.

With that said, I guess one middle-way solution would be, as you suggested:

* Remove `order`
* Do NOT remove `default_order`
* Extend `as_type` to also accept an order parameter. E.g. `vec.as_type(order='F') in vec.space` is still true.

I guess this should handle the problems you have, while also allowing copy-less implementations.

kohr-h commented:
I see your point, and I'm cool with the suggestion. Regarding interfacing, the alternative (which this is heading to) is to not give any guarantees on a space level (and clearly communicate that lack of guarantee) but instead create methods that help enforcing contiguousness in some ordering when needed. This is closer to the Numpy approach of not having an `order` property of arrays but instead just track properties of arrays and provide methods that make a contiguous copy if necessary.

So I'll go ahead and implement this, suggesting the following methods for amendment with an `order` option:
- `space.element()`
- `space.zero()`
- `space.one()`
- `vec.asarray()`
- `vec.copy()`
- `vec.astype()` (how about a different name here, e.g. just `as`? `vec.as(dtype=float)`?

The default will then be `default_order`.

adler-j commented:
My suggestion is that we chill on adding `'order'` across the board for now. Adding them is an incremental change that doesn't break anything else. Simply fix the "global" stuff and try to get this merged.

kohr-h commented:
> My suggestion is that we chill on adding 'order' across the board for now. Adding them is an incremental change that doesn't break anything else. Simply fix the "global" stuff and try to get this merged.

Okay, I'll try to restrain myself. But it will already be needed in `element` since otherwise creation from data pointer won't work.

adler-j commented:
Sounds reasonable!

kohr-h commented:
All the contrib stuff seems to work just fine, except for the tensorflow integration, which is quite important. @adler-j the errors I get from tf are cryptic, you probably know the places where changes are necessary. In most cases it's `shape` vs. `size`.

Otherwise, I'll move out the testing infrastructure changes, then we're good to go I think.

kohr-h commented:
Apart from the issue with tensorflow, this is ready to go in from my perspective. The only significant change since last review is the removal of `order` from space, this now works smoothly. It's all in one commit: b1dfcbb9.

kohr-h commented:
Not sure if you can put up with another review @adler-j 😁

kohr-h commented on odl/discr/discretization.py:
```diff
@@ -141,7 +140,7 @@ def interpolation(self):
             raise NotImplementedError('no interpolation operator provided')
 
-    def element(self, inp=None, **kwargs):
+    def element(self, inp=None, order=None, **kwargs):
```
Forgot to document `order` here

kohr-h commented on odl/discr/lp_discr.py:
```diff
@@ -269,32 +276,45 @@ def is_uniformly_weighted(self):
+            - ``None``: an empty element is created with no guarantee of
+              its state (memory allocation only). The new element will
+              use ``order`` as storage order if provided, otherwise
+              `default_order`
```
dot

adler-j replied:
yes!

kohr-h commented on odl/discr/lp_discr.py:
```diff
@@ -269,32 +276,45 @@ def is_uniformly_weighted(self):
+            - callable: a new element is created by sampling the function
+              using the `sampling` operator.
+
+        order : {'C', 'F'}, optional
```
Add `None` to the list.

adler-j replied:
> Add None to the list.

kohr-h commented on doc/source/dev/extend.rst:
```diff
@@ -10,18 +10,28 @@ This lowers the requirement on code maturity, completeness of documentation, uni
+These are in general created through the `rn` and `uniform_discr` functions who take an ``impl`` parameter, allowing users to select the backend for array storage and computations.
+
+In the core ODL package, there is only a single backend available: `NumpyTensorSpace`, given by ``impl='numpy'``, which is the default choice.
+Users can add CUDA support by installing the add-on library `odlcuda`_, which contains the additional space ``CudaFn``.
```
Maybe I should remove that for now.

kohr-h commented:
I'm gonna add a maintenance commit for imports, comments and that kind of stuff. It introduces quite a bit of noise. So you may want to exclude it from the comparison. Here's a link, you can also adjust the base commit so you don't get the full diff.

https://github.com/odlgroup/odl/compare/master...9988b49

adler-j commented:
Finally gonna get around to do this review. Could you give the exact commits that need a review, e.g. from - to?

kohr-h commented:
Sure. I'd say the earliest checkpoint would be 3a6387a, which is a bit before I started all the `__array_ufunc__` business. If you prefer to start after that (since you have already seen it), a good starting point would be 3bf1ad2. The end point I suggest would be 9988b49, that's just before a boring clean-up commit. You can have a look at that in isolation if you want, it's 08d3d48.

kohr-h commented:
I think I wanted to check something regarding boolean or index array indexing. I'll add it to the TODO list. Also, in the light of #1196 I think I'll make a PR on master for the changes to string representations, so this one becomes a bit slimmer.

Are you reviewing yet @adler-j? If not, perhaps wait until after the weekend so I can break out those changes.

adler-j commented:
Not yet reviewing, go ahead.

adler-j commented on odl/discr/lp_discr.py:
```diff
@@ -228,22 +222,29 @@ def meshgrid(self):
-        order : {None, 'C', 'F'}
-            Axis ordering in the resulting point array. For the default
-            ``None``, `default_order` is used.
+        order : {'C', 'F'}
```
Shouldn't None be default_order?

kohr-h replied:
No, that will destroy views. `default_order` only applies to new elements.

adler-j commented on odl/discr/lp_discr.py:
```diff
@@ -339,28 +352,31 @@ def element(self, inp=None, **kwargs):
+            # fspace element -> discretize
+            inp_elem = self.fspace.element(inp, vectorized=vectorized)
+            sampled = self.sampling(inp_elem, **kwargs)
+            return self.element_type(self, self.tspace.element(sampled,
```
break line after `self,`, easier to read

adler-j commented on odl/discr/lp_discr.py:
```diff
@@ -449,15 +465,11 @@ def __getitem__(self, indices):
+                fspace = space.fspace.byaxis_in[indices]
                 part = space.partition.byaxis[indices]
 
-                if isinstance(space.weighting, NoWeighting):
```
Where did this go? Overall NoWeighting is different from ConstWeighting (conceptually at least), For example, if we make a ray transform with domain `odl.uniform_discr([-1, -1], [1, 1], [2, 2])`, we still want the range to be weighted, even though the domain has weighting 1 and is thus in some sense equivalent to NoWeighting.

kohr-h replied:
I don't see how the weightings of domain and range are connected for ray transforms. They're totally different spaces.

The only situation where weighting vs. no weighting makes a difference is when the cell sizes are changed from 1 to something else, or the other way around; plus the case when axes are removed or added.
The only place where we change cell sizes is `uniform_discr_fromdiscr`, I'll have to check what we do with weightings there.
Removing axes will necessarily destroy weightings, unless there's a way to regenerate the lower-dimensional versions of them. This possibility exists in `DiscreteLp`, but not in `TensorSpace`. Later on, when we implement per-axis weighting, that will be preserved, too.

adler-j commented on odl/operator/tensor_ops.py:
```diff
@@ -844,59 +879,646 @@ def inverse(self):
+        >>> op = odl.SamplingOperator(space, sampling_points)
+        >>> x = space.element([[1, 2, 3],
+        ...                    [4, 5, 6]])
+        >>> op.adjoint(op(x)).inner(x) - op(x).inner(op(x)) < 1e-10
```
This still holds

adler-j commented on odl/phantom/noise.py:
```diff
@@ -225,7 +225,7 @@ def salt_pepper_noise(vector, fraction=0.05, salt_vs_pepper=0.5,
         else:
             # Extract vector of values
-            values = as_flat_array(vector).copy()
+            values = vector.asarray().flatten()
```
Flatten is really slow and always copies, do we want that?

kohr-h replied:
Here we explicitly want a copy at the beginning.

adler-j commented on odl/space/base_tensors.py:
```diff
@@ -444,25 +424,49 @@ def examples(self):
+
+        Returns
+        -------
+        zero : Tensor
```
Ticks on "Tensor"

adler-j commented on odl/space/base_tensors.py:
```diff
@@ -866,27 +759,37 @@ def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
         # --- Evaluate ufunc --- #
 
-        # Trivial context to be used when an output is `None`
+        # Trivial context used to create a single code path for the ufunc
```
This should be moved to `odl.util`

kohr-h replied:
I have my doubts. Where would this ever be useful?

adler-j commented on odl/space/fspace.py:
```diff
@@ -1323,8 +1330,11 @@ def __call__(self, x, out=None, **kwargs):
+                    # scalar case. The resulting array may be read-only,
+                    # in which case we copy.
                     out = broadcast_to(out, out_shape)
+                    if not out.flags.writeable:
```
Are there tests in place for corner cases like this?

kohr-h replied:
Now yes :-)
This is not a corner case, it applies whenever broadcasting happens.

adler-j commented on odl/space/npy_tensors.py:
```diff
@@ -283,17 +283,22 @@ def __init__(self, shape, dtype=None, order='A', **kwargs):
-        return self.weighting.exponent
+    def default_order(self):
+        """Default storage order for new elements in this space: ``'C'``."""
+        return 'C'
```
no way to select this?

kohr-h replied:
I cannot find an obvious way, but I think it's safe to say that Numpy will use C ordering by default for a while.

adler-j commented on odl/space/npy_tensors.py:
```diff
@@ -302,10 +307,17 @@ def weighting(self):
     @property
     def is_weighted(self):
-        """Return ``True`` if the space has a non-trivial weighting."""
-        return not isinstance(self.weighting, NumpyTensorSpaceNoWeighting)
```
See earlier comment on the removal of NoWeighting, it is not "ok" so to say.

adler-j commented on odl/space/npy_tensors.py:
```diff
@@ -1774,6 +1795,22 @@ def fallback_copy(x1, x2, n):
+        if out.data.flags.f_contiguous:
+            ravel_order = 'F'
+        else:
+            ravel_order = 'C'
```
What if the array is not contiguous in any sense?

kohr-h replied:
Then we use C order. The F order thing is just for performance.
We need to flatten all arrays in the same fixed order, since otherwise the results will depend on the axis ordering of the inputs. I've had that bug :-)

adler-j commented on odl/test/discr/lp_discr_test.py:
```diff
@@ -769,13 +776,14 @@ def test_ufuncs(tspace_impl, ufunc):
-        # In-place will not work with Numpy < 1.13
-        odl_result_new = elem_fun_new(*in_elems_new, **out_elem_kwargs)
-        assert all_almost_equal(npy_result, odl_result_new)
+    with np.errstate(all='ignore'):
```
add a comment on this ignore

adler-j commented on odl/test/largescale/space/tensor_space_slow_test.py:
```diff
@@ -40,14 +40,15 @@ def test_element(tspace):
 
-    y = tspace.element(inp=[0] * tspace.size)
+    # From array-like
+    y = tspace.element(np.zeros(tspace.shape, dtype=tspace.dtype).tolist())
```
why tolist?

kohr-h replied:
To get an array-like

adler-j commented on odl/test/largescale/tomo/analytic_slow_test.py:
```diff
@@ -174,7 +174,8 @@ def test_fbp_reconstruction(projector):
 
-    maxerr = vol.norm() / 5.0
+    # Allow 30 % error
+    maxerr = vol.norm() * 0.3
```
why increase?

kohr-h replied:
I'll have to check that again, it sometimes failed with the lower error margin (also on master).

kohr-h commented:
I now squashed everything into one commit that I'll split into smaller packages (and fix apparently). Too hard to rebase otherwise.

adler-j commented:
Please go ahead, I'll help you out if there is anything!

kohr-h commented on examples/trafos/wavelet_trafo.py:
```diff
@@ -20,7 +20,7 @@
 # It may however (for some choices of wbasis) be interpreted as a vector in the
 # domain of the transformation
-phantom_wt_2d = space.element(phantom_wt)
+phantom_wt_2d = space.element(phantom_wt.asarray().reshape(space.shape))
```
No, it will try to `__array_wrap__` it and fail.

kohr-h commented:
Okay, all the things I had on my list are fixed now. The only thing that remains is the fix for the TF bindings.

kohr-h commented:
I figured out what to do with the TF bindings. The issue was only affecting the product spaces. Fixed now.

adler-j commented:
Does this mean that we're "go"? :O

kohr-h commented:
I pretty much think so. I'm satisfied with the current state!

adler-j commented:
One question I have is:

Does

```
productspace_el.asarray().shape == productspace.shape?
```

kohr-h commented:
> One question I have is:
> 
> Does
> 
> productspace_el.asarray().shape == productspace.shape?

No, because `productspace_el.asarray()` is still not implemented. That was a separate issue.

adler-j commented:
Well once it is implemented, wouldn't we expect that to hold? I guess it would be non-breaking though

kohr-h commented:
> Well once it is implemented, wouldn't we expect that to hold? I guess it would be non-breaking though

We can still fix that on master, there's no harm in leaving it unimplemented.

adler-j commented:
I guess we're go then!

kohr-h commented:
Alright, big moment!!

kohr-h commented:
BAM!

adler-j commented:
Way more issues should be closed by this :D

kohr-h commented:
I'm looking through the list.

On 11/13/2017 06:27 PM, Jonas Adler wrote:
> Way more issues should be closed by this :D
> 

-- 

Holger Kohr
Centrum Wiskunde & Informatica
Computational Imaging
Science Park 123
1098 XG Amsterdam
